File: HW10/a00824566.hs
-- ====================
-- Complete the following functions and submit your file to Canvas.
-- ====================
-- Do not change the names of the functions. 
-- Do not change the number of arguments in the functions.
-- If your file cannot be loaded by the Haskell compiler, your submission may be cancelled. 
-- Then, submit only code that works.
-- ====================
-- Grading instructions:
-- There is a series of test cases for each function. In order to state that your function
-- "works as described", your output must be similar to the expected one in each case.

-- === Distance between two points ===

data Point = Point Double Double deriving Show

distance :: Point -> Point -> Double
distance (Point x y) (Point a b) = sqrt((x - a) ** 2 + (y - b) ** 2)

-- === Complex numbers ===

data Complex = Complex Int Int deriving Show

add :: Complex -> Complex -> Complex
add (Complex x y) (Complex a b) = (Complex (x + a) (b + y))

multiply :: Complex -> Int -> Complex
multiply (Complex x y) c = (Complex (c * x) y)

getReal :: Complex -> Int
getReal (Complex x _) = x

getImaginary :: Complex -> Int
getImaginary (Complex _ y) = y

-- === The RGB color mode ===

-- Yes, the components are given as a tuple inside the data type...
data RGB = RGB (Int, Int, Int) deriving Show

getR :: RGB -> Int
getR (RGB (r, _, _)) = r

getG :: RGB -> Int
getG (RGB (_, g, _)) = g

getB :: RGB -> Int
getB (RGB (_, _, b)) = b

getMaxComponent :: RGB -> ([Char], Int)
getMaxComponent (RGB (r, g ,b)) = if r > g
 then if r > b
 then ("Red", r)
 else ("Blue", b)
 else if g > b
 then ("Green", g)
 else ("Blue", b)

combine :: RGB -> RGB -> RGB
combine (RGB (r1, g1, b1)) (RGB (r2, g2, b2)) = (RGB (div (r1 + r2) 2, div (g1 + g2) 2, div (b1 + b2) 2))

-- === Playing cards ===

data Suit = Diamonds | Clubs | Hearts | Spades deriving (Eq, Show)
data Color = Red | Black deriving Show
data Rank = Jack | Queen | King | Ace | Two | Three | Four | Five | Six | Seven | Eight | Nine | Ten deriving (Eq, Show)
data Card = Card Rank Suit deriving Show

getValue :: Card -> Int
getValue (Card x _) = case x of
 Two -> 2
 Three -> 3
 Four -> 4
 Five -> 5
 Six -> 6
 Seven -> 7
 Eight -> 8
 Nine -> 9
 Ten -> 10
 Jack -> 11
 Queen -> 12
 King -> 13
 Ace -> 14
 
getColor :: Card -> Color
getColor (Card _ x) = if x == Hearts || x == Diamonds
 then Red
 else Black

-- === Test cases ===

main = do 
    print "=== Distance between two points ==="
    print "=== distance ==="
    print $ distance (Point 10 20) (Point 5 15) -- 7.0710678118654755
    print "=== Complex numbers ==="
    print "=== add ==="
    print $ add (Complex 5 6) (Complex 10 2) -- Complex 15 8
    print "=== multiply ==="
    print $ multiply (Complex 3 4) 6 -- Complex 18 4
    print "=== getReal ==="
    print $ getReal (Complex 2 3) -- 2
    print "=== getImaginary ==="
    print $ getImaginary (Complex 2 3) -- 3
    print "=== The RGB color mode ==="
    print "=== getR ==="
    print $ getR (RGB (150, 30, 25)) -- 150
    print "=== getG ==="
    print $ getG (RGB (150, 30, 25)) -- 30
    print "=== getB ==="
    print $ getB (RGB (150, 30, 25)) -- 25
    print "=== getMaxComponent ==="
    print $ getMaxComponent (RGB (17, 78, 63)) -- ("Green", 78)
    print "=== combine ==="
    print $ combine (RGB (120, 120, 60)) (RGB (150, 70, 80)) -- RGB (135, 95, 70)
    print "=== Playing cards ==="
    print "=== getValue ==="
    print $ getValue (Card King Diamonds) -- 13
    print $ getValue (Card Three Spades) -- 3
    print "=== getColor ==="
    print $ getColor (Card King Diamonds) -- Red
    print $ getColor (Card Three Spades) -- Black
    

File: EX03/a00824566.pl
% ====================
% Complete the following programs and submit your file to Canvas.
% ====================
% Do not change the names of the programs. 
% Do not change the number of elements in the clauses.
% If your file cannot be loaded by the Prolog interpreter, your submission may be cancelled. 
% Then, submit only code that works.
% ====================
% Grading instructions:
% There is a series of test cases for each program. In order to state that your program
% "works as described", your output must be similar to the expected one in each case.

:- use_module(library(bounds)).

% === tea ===

tea(P, E, A, T, X) :-
	between(0, 100, P),
  between(0, 100, E),
  between(0, 100, A),
  between(0, 100, T),
  all_different([P, E, A, T]),
  22 is P + E + A,
  42 is P + E + A + T,
  X is T + E + A, !.

% === candy ===

candy(Sa, Su, J) :-
	Sa is 15,
  Su is Sa + (Sa * 2),
  X is ((Sa + Su) div 5) * 8,
  J is (X div 8) * 3, !.

% === ages ===

ages(Bo, Be, T) :-
  between(1, 100, Bo),
  between(1, 100, Be),
  between(1, 100, T),
  T is Bo + Be,
  T is (Bo**3 + Be**3) div T,
  T is (Bo**2 + Be**3) div T, !.

% === houses ===

right(X, Y, L) :- append(_, [X, Y|_], L).
next(X, Y, L):- right(X, Y, L) ; right(Y, X, L).
before(X, Y, Solution) :- nth1(X1, Solution, X), nth1(Y1, Solution, Y),	X1 < Y1.
after(X, Y, Solution) :- before(Y, X, Solution).

houses(Solution) :-
  length(Solution, 4),
  member([_, red], Solution),
  member([_, yellow], Solution),
  member([_, blue], Solution),
  member([_, green], Solution),
  Solution = [_, _, [bruce, _], _],
  member([tony, green], Solution),
  next([peter, _], [bruce, _], Solution),
  \+member([steve, blue], Solution),
  before([steve, _], [peter, _], Solution), 
  (Solution = [[_, red], _, _, _]; Solution = [_, _, _, [_, red]]),
  member([steve, _], Solution), !.

start :-
	write("% === tea ==="), nl, 
	tea(P, E, A, T, X),
  Tea is T + E + A,
	write(Tea), nl,
  write("% === candy ==="), nl, 
	candy(Sanjev, Susan, Jack),
	write(Jack), nl, 
  write("% === ages ==="), nl, 
	ages(Bob, Ben, Tom),
  Ages = [["Bob", Bob], ["Ben", Ben], ["Tom", Tom]],
	write(Ages), nl,
  write("% === houses ==="), nl, 
	houses(Solution),
	write(Solution), !.